Introduces asynchronous logging

tinwelint opened the pull request.

chrisvest committed:
AsyncEvents concurrency primitive

The AsyncEvents concurrency primitive can be used for very quickly enqueueing events to be processed in a background thread.

This will be useful for building asynchronous logging.

tinwelint committed:
Introduces asynchronous logging

Where AsyncLogProvider wraps an actual LogProvider, and internally all sub
components thereof, providing the asynchronous aspect of the logging.
Each logging statement is registered as an event, to be processed by
a separate and dedicated logging thread in the background.

The driving reason for this is to avoid I/O waits on critical execution
paths that does, even ever so little, logging.

Implementation uses AsyncEvents, which is a lock-free unbounded queue with
a consumer for processing the log events, i.e. making the actual logging
happen. Logging in neo4j is considered to be scarse enough to not impose a
heap usage problem which may theoretically be a problem for an unbounded
queue.

srbaker commented:
~1600 additions, across 25 files, and only two commits. Yikes.

srbaker commented:
A quick search reveals that log4j provides asynchronous logging services. What does our solution do that it doesn't? What do we get from the maintenance overhead?

chrisvest commented:
Tests fail because they run out of heap. They most likely pile up too many log messages because the background log thread hasn't been started. We can fix this by making AsyncEvents process the events in the sending thread when the background thread hasn't been started: If we initialise the stack to null instead of the end node, the send method can start out reading the stack, and if it observes a null, process the event directly. Then return early, before the getAndSet.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,134 @@
+                    // What's most important is that logging gets going, but we also can be nice and...
+                    // well... not log, since apparently there was a problem doing that right now, but do something.
+                    // That something will be to print stack trace of the exception, capped by different
+                    // exception types and time interval for good measure.
```
this comment looks not completed

tinwelint replied:
@davidegrohmann What's missing?

davidegrohmann replied:
nevermind, my mistake

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,134 @@
+            @Override
+            public void receive( Throwable error )
+            {
+                error.printStackTrace();
```
This doesn't look like a good idea

tinwelint replied:
The comment above (or below in the code) tries to outline the scenario leading up to this point. Is it a good idea? Don't know. What can we do instead?

tinwelint commented:
@chrisvest memory was piling up since we added the LogService twice to the db lifecycle, which in end caused one stray background Thread for logging to linger, which in extension had a reference to the db. And so all databases every test every started was left in heap. Fixed now.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,134 @@
+    @Override
+    public Log getLog( @SuppressWarnings( "rawtypes" ) Class loggingClass )
+    {
+        return new AsyncLog( events, actual.getLog( loggingClass ) );
```
It is probably a better idea to cache here `AsyncLog`s rather than in `AbstractLogProvider`.

tinwelint commented:
@srbaker what do we get with our own logging framework? One argument is less dependencies... there were apparently wide-spread problems with our logback dependency we had in 2.2, conflicting with a different version of logback that the user would already have. Same goes for log4j.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/lifecycle/LifeSupport.java:
```diff
@@ -307,6 +307,8 @@ public synchronized void restart( Lifecycle instance )
     {
         if ( instance instanceof Lifecycle )
         {
+            checkContains( (Lifecycle) instance );
```
I love this, this should be a great help in finding issues around lifecycle management.

Since how we wire things up is mostly static though, perhaps we should only do this when assertions are enabled?

davidegrohmann replied:
:+1:

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/StoreLogService.java:
```diff
@@ -149,40 +167,81 @@ private StoreLogService( LogProvider userLogProvider,
+                    builder.asyncExecutor.execute( asyncInternalLogProvider );
+                }
+            };
+            closeInShutdown = new MultiCloseable( asyncUserLogProvider, asyncInternalLogProvider, closeInShutdown );
```
It looks like you want to use `LifeCycle` here instead...

thobe replied:
well... it does use `Lifecycle`, since the `StoreLogService` implements that interface.

tinwelint replied:
@davidegrohmann You're right. Converting all that jazz into Lifecycle (and having an internal LifeSupport) removed loads of stuff and was a very good fit. Take a look again.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/MultiCloseable.java:
```diff
@@ -0,0 +1,104 @@
+        {
+            throw launderedException( IOException.class, exception );
+        }
+    }
```
Why not using `LifeCycle` for this? I don't see the point of adding so much extra complexity since we have already code that does something like this.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+{
+    // TODO use VarHandles in Java 9
+    private static AtomicReferenceFieldUpdater<AsyncEvents,AsyncEvent> updater =
+            AtomicReferenceFieldUpdater.newUpdater( AsyncEvents.class, AsyncEvent.class, "stack" );
```
I prefer naming my updaters the same name as the field they update (although using uppercase for the updater, and lowercase for the volatile filed), I find it typically makes the code clearer to read.

chrisvest replied:
Sounds like a good naming habit.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/logging/AsyncLogProviderTest.java:
```diff
@@ -0,0 +1,198 @@
+        waitForLogMessage( actualLogProvider, "After error" );
+        assertFalse( actualLogProvider.hasLogged( "Should throw" ) );
+        verify( errorHandler, times( 1 ) ).receive( any( Throwable.class ) );
+    }
```
What about testing timestamps? I would expect them to be the event creation time and not the logging time...

tinwelint replied:
That's probably what we want, but not really possible w/o some refactorings in the logging framework. What we could do is to note the timestamp for when the event was created and compare with when it's logged. If there's a > threshold difference there we can add an additional log statement saying something like `BLOCKED for X ms`.

thobe replied:
Oh yes, I definitely agree, the timestamp should be the time of event creation.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogger.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public void log( String message )
+    {
+        sender.send( logEvent( logger, message ) );
```
we should be aware that this potentially will create a lot of garbage...

tinwelint replied:
Obviously we should measure, but I don't think this is a problem since we don't have very much logging. The garbage of the event instances are dwarfed by the strings and exceptions themselves.

davidegrohmann replied:
Yep, we should measure.

davidegrohmann commented on community/logging/src/main/java/org/neo4j/logging/BufferingLog.java:
```diff
@@ -400,6 +444,7 @@ public void replayInto( Log other )
             while ( message != null )
             {
                 message.replayInto( other );
+                System.out.println( "Replaying message " + message );
```
to be removed

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+
+    @SuppressWarnings( {"unused", "FieldCanBeLocal"} )
+    private volatile AsyncEvent stack; // Accessed via AtomicReferenceFieldUpdater
+    private volatile boolean shutdown;
```
Should we add padding between the various fields of this object to eliminate false sharing?

chrisvest replied:
The `shutdown` field is hardly ever written to, and only read from the background thread. I'm not sure it's worth the cost in code aesthetics.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+
+    private static final class EndEvent extends AsyncEvent
+    {
+        volatile Thread runner;
```
Why is the running thread part of the end-sentinel? It seems more intuitive to have it be part of the `AsyncEvents` object itself.

If we did that the end sentinel could turn into a constant.

chrisvest replied:
+1

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+            AtomicReferenceFieldUpdater.newUpdater( AsyncEvents.class, AsyncEvent.class, "stack" );
+
+    private final EndEvent endSentinel;
+    private final ShutdownEvent shutdownSentinel;
```
The shutdown sentinel could be turned into a constant. And it might be possible to do the same with the endSentinel.

chrisvest replied:
Yes, both can be constants if we move the `runner` field from the `EndEvent` class to the `AsyncEvents` class.

davidegrohmann commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+
+    private static final class ShutdownEvent extends AsyncEvent
+    {
+    }
```
The code in `AsyncEvents` feels overly too complicated for no obvious reason.

tinwelint replied:
It's a technique already in use in `WorkSync` so it's not new. WorkSync is used in 2.2 and 2.3 during the commit process. We could unify them though.

davidegrohmann replied:
Perhaps, but that doesn't answer why we need that solution here as well.

davidegrohmann commented:
@tinwelint @chrisvest isn't it better to adopt some already existent logging framework? They are definitely more robust and tested than our own adhoc solution.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,134 @@
+            };
+
+            @Override
+            public void accept( AsyncLogEvent logEvent )
```
Wouldn't it be a lot better if we could get a whole batch of log events here, so that we can do a single batch of writes to the underlying logging system, thus limiting the number of IO-operations we perform...

tinwelint replied:
I think we should turn off flushing on every log statement in the logging framework, when we do async logging, and instead flush after each batch (batch in the sense of each time the async thread awakes and processes all pending log messages).

thobe replied:
yes, that is one way of achieving that.

tinwelint commented:
@davidegrohmann about logging framework: see my reply to Steven Baker above.

davidegrohmann commented:
@tinwelint well I disagree that we should go with our own logging framework for only minimizing dependencies. I would prefer to ship with a more robust logging framework and deal with the extra dependency.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/StoreLogService.java:
```diff
@@ -149,40 +167,81 @@ private StoreLogService( LogProvider userLogProvider,
+                public void run()
+                {
+                    builder.asyncExecutor.execute( asyncUserLogProvider );
+                    builder.asyncExecutor.execute( asyncInternalLogProvider );
```
This looks like we will get two different log processing threads, one for each of the two logs we have.
Since we (most likely) only have one I/O subsystem, would it not be better to process both event streams on the same thread?

That would mean enqueueing the events for both logs into the same `AsyncEvents` queue (since the events know which log they are writing to already, this should be fine).

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLog.java:
```diff
@@ -0,0 +1,84 @@
+ * Wraps a {@link Log}, making its logging asynchronous by letting log statements be treated as events fed into a
+ * {@link AsyncEventSender} to be processed later instead of in the thread making the log statement call.
+ */
+class AsyncLog extends AbstractLog
```
The way this is implemented, we still enqueue the logging events, even if the log-level is not enabled.
That is quite wasteful (especially since the finer log-levels are the most likely to not be enabled, and also the most verbose ones).

tinwelint replied:
Fixed now

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLog.java:
```diff
@@ -0,0 +1,84 @@
+        this.infoLogger = new AsyncLogger( sender, actual.infoLogger() );
+        this.warnLogger = new AsyncLogger( sender, actual.warnLogger() );
+        this.errorLogger = new AsyncLogger( sender, actual.errorLogger() );
+        this.actual = actual;
```
By accessing the loggers for the different levels here, we eliminate the ability to change the log-levels later.

If for example debug-logging is disabled at start up, the logger we get from `actual.debugLogger()` will be a null-logger. If debug-logging is then enabled later, this class will not update its reference to that underlying logger, effectively making it impossible to change log levels.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/lifecycle/LifeSupport.java:
```diff
@@ -307,6 +307,8 @@ public synchronized void restart( Lifecycle instance )
         if ( instance instanceof Lifecycle )
         {
+            assert notAlreadyAdded( (Lifecycle) instance );
+
```
I really think this change should go in `2.2` as well.

tinwelint replied:
That makes sense. We can backport later.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,142 @@
+                    }
+                    catch ( Throwable e )
+                    {
+                        // Even that failed. There's no reasonable thing left to do here, is there?
```
at least print stacktrace?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,142 @@
+    AsyncLogProvider( LogProvider actual, Executor scheduler, final Listener<Throwable> errorHandler )
+    {
+        this.scheduler = scheduler;
+        this.events = new AsyncEvents<>( new Consumer<AsyncLogEvent>()
```
Why do we introduce new structure for AsyncEvents? How performance wise it's comparable with different kinds of j.u.c.\* Queues?

thobe commented on community/logging/src/main/java/org/neo4j/logging/Log.java:
```diff
@@ -144,6 +144,15 @@
+     * @param level {@link Level} of logging to get a {@link Logger} for.
+     * @return a {@link Logger} for specified {@code level}.
+     */
+    Logger logger( Level level );
```
This is only used in a test. Is it really needed?

tinwelint replied:
Perhaps not

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/StoreLogService.java:
```diff
@@ -149,56 +167,81 @@ private StoreLogService( LogProvider userLogProvider,
+            // where the scheduling into the executor takes place during init().
+            // Also Reassign these log providers so that the async versions are used below.
+            userLogProvider = internalLife.add( new AsyncLogProvider( userLogProvider, builder.asyncExecutor ) );
+            internalLogProvider = internalLife.add( new AsyncLogProvider( internalLogProvider, builder.asyncExecutor ) );
```
This looks like we will get two different log processing threads, one for each of the two logs we have.
Since we (most likely) only have one I/O subsystem, would it not be better to process both event streams on the same thread?

That would mean enqueueing the events for both logs into the same AsyncEvents queue (since the events know which log they are writing to already, this should be fine).

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+        shutdownSentinel.next = null;
+    }
+
+    private AsyncEvent reverseAndStripEndMark( AsyncEvent events )
```
why do we use stack and doing reverse instead of deque afor example?

thobe replied:
Because a stack implies fewer pointers to maintain invariants on, making the code for adding events simpler and faster.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java:
```diff
@@ -0,0 +1,142 @@
+        // Flush the queue right after we've shut it down since we don't want to have pending logging
+        // events after leaving this method. Right after leaving this method the actual output stream
+        // will most likely be closed.
+        events.run();
```
Imho this run should be part of `events` itself and not to be leaked to AsyncLogProvider.
Events should insure that all queued messages are flushed on shutdown, instead of AsyncLogProvider doing that.

thobe replied:
Very similar to the issue I pointed out here: https://github.com/neo4j/neo4j/pull/5599#discussion_r41745847 - two sides of the same coin.

thobe commented:
On a grander scheme of things, I agree with most other reviewers that it would probably be a safer bet to use an existing logging framework that comes with solutions for the issues we are trying to solve out of the box.
If the argument against that is that it would conflict with other versions of that same framework in embedded applications, we should at least attempt to [shade](https://maven.apache.org/plugins/maven-shade-plugin/) that library before jumping to the conclusion that we need to make our own very sophisticated home grown framework.

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,162 @@
+    public void shutdown()
+    {
+        shutdown = true;
+        LockSupport.unpark( endSentinel.runner );
```
On shutdown we will unpark executor thread from pool, while later on we will call events.run from current thread, having 2 threads running events as result

chrisvest replied:
Yes, the consumer has to be thread-safe. This should be mentioned in the javadocs.

MishaDemianenko commented:
Taking into account how many comments are already here and async bugs of log4j and their performance investigations/evolution (it was already mentioned here as example so just to follow) do not really think that doing yet another logging wheel is a good idea.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/logging/AsyncLogProviderTest.java:
```diff
@@ -0,0 +1,219 @@
+        assertEquals( 0, actualLog.toString().length() );
+
+        // WHEN
+//        flush( asyncLogProvider );
```
What's this about?

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java:
```diff
@@ -0,0 +1,175 @@
+    @SuppressWarnings( {"unused", "FieldCanBeLocal"} )
+    private volatile AsyncEvent stack; // Accessed via AtomicReferenceFieldUpdater
+    private volatile Thread runner;
+    private volatile boolean shutdown;
```
@chrisvest the main padding I'd like to add is between the `eventConsumer` and the `stack`, so that we don't need to re-fetch the cache line to access the `eventConsumer`.
After that I don't really care which of those groups `runner` and `shutdown` should go into (but I think these two should go together).

thobe replied:
@chrisvest: I was thinking something like this: https://github.com/tinwelint/neo4j/pull/5

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/lifecycle/Lifecycles.java:
```diff
@@ -66,4 +66,26 @@ public void shutdown() throws Throwable
+        @Override
+        public void shutdown() throws Exception
+        {
+            System.out.println( "Shutting down " + closeable );
```
Oops!

thobe commented:
The "Only adds LogService to LifeSupport once" commit should be excluded now. It has been merged through 2.2 from #5552

tinwelint commented:
@MishaDemianenko at this point we're trading stability of logging for stability of clustering, which to me is given who wins. I'd much prefer a stable cluster and fixing bugs related to asynchronous logging later than the other way around.

MishaDemianenko commented:
@tinwelint at this point fully agree, we can schedule switch to more robust logging framework for later as well, for now i would be more then happy if we will agree that we need to do that switch on some point in nearest future

davidegrohmann commented:
@tinwelint @MishaDemianenko I disagree, moving bugs around the code base is not a solution

tinwelint commented:
@davidegrohmann I don't understand what you mean. Surely some things are more important than others. Moving bugs around is part of everyone's day to day job, ain't it ;)

tinwelint commented:
@MishaDemianenko I think it's fair to have a plan to move to an established logging framework after 2.3. At least invest some time into figuring out impact of doing so.

MishaDemianenko commented:
@davidegrohmann fully agree with you that it's a really wrong way in the same time keeping in mind current deadlines and choosing between two evils, would agree with @tinwelint  that stable cluster is more important.

davidegrohmann commented:
@tinwelint I meant that it is not acceptable solution to have a problematic logging framework even though it will make sure that clustering is stable.

chrisvest committed:
Make AsyncEvents.awaitTermination use a latch

Instead of spinning on `Thread.sleep`. Also add a test for it.

cleishm commented:
What is the problem we're trying to solve? What are the problems in the current logging?

I don't think jumping to pull in yet-another-logging-framework is a good idea. Our needs are extremely simple, and I haven't seen any issue with the very simple approach we have.

davidegrohmann commented:
@cleishm we have tons of issues in clustering with the current logging solution, it effects both performance and stability

cleishm commented:
@davidegrohmann How can logging possibly affect the stability of clustering? That would suggest that clustering is extremely sensitive to timing, which, if true, should probably be the real issue to fix.

thobe commented:
@cleishm I wouldn't say clustering is _extremely_ sensitive to timing. But it is at least sensitive to timing, and logging is _extremely_ slow (when IO load is high).

tinwelint commented:
This will not be done now, perhaps not globally like this ever. So closing.
